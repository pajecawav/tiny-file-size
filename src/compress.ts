import zlib from "zlib";

export function getGzipSize(buffer: Buffer): Promise<number> {
	return new Promise((resolve, reject) => {
		zlib.gzip(buffer, (error, result) => {
			if (error) return reject(error);
			resolve(result.length);
		});
	});
}

export function getBrotliSize(buffer: Buffer): Promise<number> {
	return new Promise((resolve, reject) => {
		zlib.brotliCompress(buffer, (error, result) => {
			if (error) return reject(error);
			resolve(result.length);
		});
	});
}
