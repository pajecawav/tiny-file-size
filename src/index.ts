export { getBrotliSize, getGzipSize, getRawSize } from "./compress";
export { buildJsonReport, buildPrettyReport, humanizeBytes } from "./report";
export type { PrettyReportOptions } from "./report";
export { getFileSizes, run } from "./run";
export type { FileSize, FileSizesOptions } from "./run";
