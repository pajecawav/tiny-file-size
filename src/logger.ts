import colors from "picocolors";

export const logger = {
	error(message: string) {
		console.error(colors.bgRed(colors.bold(colors.black(" ERROR "))), colors.red(message));
	},
};
