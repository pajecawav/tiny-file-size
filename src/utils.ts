export function sum(values: number[]): number {
	let total = 0;
	for (const value of values) total += value;
	return total;
}
