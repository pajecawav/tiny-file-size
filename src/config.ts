import colors from "picocolors";
import { version } from "../package.json";
import { logger } from "./logger";

export interface Config {
	globs: string[];
	gzip: boolean;
	brotli: boolean;
	output: "json" | "pretty";
	total: boolean;
}

const HELP = `
${colors.bold("Usage:")}
  file-size [options] [globs...]

${colors.bold("Options:")}
  -h, --help                     Show help
  -g, --gzip                     Include gzip size in output 
  -b, --brotli                   Include brotli size in output 
      --total                    Include total size in output
      --json                     Print result as json
  -v, --version                  Print the current version and exit

${colors.bold("Examples:")}
  # Show sizes of files foo.js and bar.json
  file-size foo.js bar.json

  # Include sizes after gzip and brotli
  file-size --gzip --brotli foo.js bar.json

  # Print output as JSON
  file-size --json foo.js bar.json
`.trim();

export function parseArgs(args: string[]): Config {
	const cfg: Config = {
		globs: [],
		gzip: false,
		brotli: false,
		output: "pretty",
		total: false,
	};

	for (let i = 0; i < args.length; i++) {
		const arg = args[i];

		if (arg[0] !== "-") {
			cfg.globs = args.slice(i);
			break;
		}

		switch (arg) {
			case "-h":
			case "--help":
				console.log(HELP);
				process.exit(0);
				break;
			case "-v":
			case "--version":
				console.log(version);
				process.exit(0);
				break;
			case "-g":
			case "--gzip":
				cfg.gzip = true;
				break;
			case "-b":
			case "--brotli":
				cfg.brotli = true;
				break;
			case "--json":
				cfg.output = "json";
				break;
			case "--total":
				cfg.total = true;
				break;
			default:
				logger.error(`Unknown argument ${arg}`);
				process.exit(1);
		}
	}

	return cfg;
}

export const config = parseArgs(process.argv.slice(2));
