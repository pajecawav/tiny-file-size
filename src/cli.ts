import { config } from "./config";
import { run } from "./run";

run(config);
