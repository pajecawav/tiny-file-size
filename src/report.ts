import path from "path";
import { bold, dim, green } from "picocolors";
import { FileSize } from "./run";

export function reportJson(sizes: FileSize[]) {
	console.log(sizes);
}

export function reportConsole(sizes: FileSize[]) {
	for (const { file, plain, gzip, brotli } of sizes) {
		const { dir, base } = path.parse(file);

		const pathStr = (dir ? dim(dir + "/") : "") + green(base);

		const plainStr = humanizeBytes(plain);
		const gzipStr = humanizeBytes(gzip);
		const brotliStr = humanizeBytes(brotli);

		console.log(`${pathStr}: ${bold(plainStr)} (${gzipStr} gzip, ${brotliStr} brotli)`);
	}
}

function humanizeBytes(bytes: number): string {
	let suffix = "MB";
	for (const suff of ["B", "KB"]) {
		if (bytes < 1024) {
			suffix = suff;
			break;
		}
		bytes /= 1024;
	}

	if (suffix === "B") {
		return `${bytes.toFixed(0)} ${suffix}`;
	}

	return `${bytes.toFixed(2)} ${suffix}`;
}
