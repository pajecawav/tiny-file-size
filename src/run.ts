import { existsSync, statSync } from "fs";
import { getBrotliSize, getGzipSize, getRawSize } from "./compress";
import { Config } from "./config";
import { logger } from "./logger";
import { buildJsonReport, buildPrettyReport } from "./report";

export interface FileSize {
	file: string;
	raw: number;
	gzip: number | null;
	brotli: number | null;
}

interface FileSizesOptions {
	files: string[];
	gzip: boolean;
	brotli: boolean;
}

export async function getFileSizes(config: FileSizesOptions): Promise<FileSize[]> {
	const sizes: FileSize[] = await Promise.all(
		config.files.map(async file => {
			const [raw, gzip, brotli] = await Promise.all([
				getRawSize(file),
				config.gzip ? getGzipSize(file) : null,
				config.brotli ? getBrotliSize(file) : null,
			]);

			return { file, raw, gzip, brotli };
		})
	);

	return sizes;
}

export async function run(config: Config) {
	for (const file of config.files) {
		if (!existsSync(file) || !statSync(file).isFile()) {
			logger.error(`No such file ${file}`);
			process.exit(1);
		}
	}

	let sizes: FileSize[];
	try {
		sizes = await getFileSizes(config);
	} catch (e: unknown) {
		if (!(e instanceof Error)) {
			throw e;
		}

		logger.error(e.message);
		process.exit(1);
	}

	let report: string;
	if (config.output === "json") {
		report = buildJsonReport(sizes);
	} else {
		report = buildPrettyReport(sizes, { total: config.total });
	}

	console.log(report);
}
