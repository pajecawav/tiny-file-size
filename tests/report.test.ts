import { describe, expect, it } from "vitest";
import { FileSize } from "../src";
import { buildPrettyReport, buildJsonReport } from "../src/report";

const sizes: FileSize[] = [
	{ file: "foo.js", raw: 123456, gzip: 5661, brotli: 123 },
	{ file: "some/sub/path/bar.md", raw: 445123, gzip: 82734, brotli: 28372 },
	{ file: "another/folder/baz.js", raw: 1234, gzip: 5678, brotli: null },
	{ file: "simple.file.ts", raw: 1234, gzip: null, brotli: 5677 },
	{ file: "test/package.json", raw: 4444, gzip: null, brotli: null },
];

describe("report", () => {
	it("json", () => {
		const report = buildJsonReport(sizes);
		expect(report).toMatchSnapshot();
	});

	it("pretty", () => {
		const report = buildPrettyReport(sizes);
		expect(report).toMatchSnapshot();
	});
});
