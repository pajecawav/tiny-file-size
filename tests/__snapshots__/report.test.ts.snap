// Vitest Snapshot v1

exports[`report > json 1`] = `
"[
    {
        \\"file\\": \\"foo.js\\",
        \\"raw\\": 123456,
        \\"gzip\\": 5661,
        \\"brotli\\": 123
    },
    {
        \\"file\\": \\"some/sub/path/bar.md\\",
        \\"raw\\": 445123,
        \\"gzip\\": 82734,
        \\"brotli\\": 28372
    },
    {
        \\"file\\": \\"another/folder/baz.js\\",
        \\"raw\\": 1234,
        \\"gzip\\": 5678,
        \\"brotli\\": null
    },
    {
        \\"file\\": \\"simple.file.ts\\",
        \\"raw\\": 1234,
        \\"gzip\\": null,
        \\"brotli\\": 5677
    },
    {
        \\"file\\": \\"test/package.json\\",
        \\"raw\\": 4444,
        \\"gzip\\": null,
        \\"brotli\\": null
    }
]"
`;

exports[`report > pretty 1`] = `
"[32mfoo.js               [39m [1m120.56 KB[22m[2m â”‚ gzip:  5.53 KB[22m[2m â”‚ brotli:    123 B[22m
[2msome/sub/path[22m/[32mbar.md [39m [1m434.69 KB[22m[2m â”‚ gzip: 80.79 KB[22m[2m â”‚ brotli: 27.71 KB[22m
[2manother/folder[22m/[32mbaz.js[39m [1m  1.21 KB[22m[2m â”‚ gzip:  5.54 KB[22m
[32msimple.file.ts       [39m [1m  1.21 KB[22m[2m â”‚ brotli:  5.54 KB[22m
[2mtest[22m/[32mpackage.json    [39m [1m  4.34 KB[22m"
`;

exports[`report > pretty with total size 1`] = `
"[32mfoo.js               [39m [1m120.56 KB[22m[2m â”‚ gzip:  5.53 KB[22m[2m â”‚ brotli:    123 B[22m
[2msome/sub/path[22m/[32mbar.md [39m [1m434.69 KB[22m[2m â”‚ gzip: 80.79 KB[22m[2m â”‚ brotli: 27.71 KB[22m
[2manother/folder[22m/[32mbaz.js[39m [1m  1.21 KB[22m[2m â”‚ gzip:  5.54 KB[22m
[32msimple.file.ts       [39m [1m  1.21 KB[22m[2m â”‚ brotli:  5.54 KB[22m
[2mtest[22m/[32mpackage.json    [39m [1m  4.34 KB[22m
[1mTotal:               [22m [1m562.00 KB[22m[2m â”‚ gzip: 91.87 KB[22m[2m â”‚ brotli: 33.37 KB[22m"
`;
